### ISSUE TITLE
Hanami 2.0 app's providers under `config/providers` are not getting prepared by default

### ISSUE THREAD

**Issue by jiggneshhgohel** [hanami/hanami#1261]
Hello,

I started exploring Hanami and as part of that started following the 2.0 guide at https://guides.hanamirb.org/v2.0/introduction/getting-started.

I am trying to access a relation class in **development env console** but am getting` NameError: uninitialized constant` related to ROM. Please find relevant code snippets below:

**Gemfile**

```
# frozen_string_literal: true

source "https://rubygems.org"

gem 'hanami', '~> 2.0'
gem 'hanami-router', '~> 2.0'
gem 'hanami-controller', '~> 2.0'
gem 'hanami-validations', '~> 2.0'

gem "dry-types", "~> 1.0", ">= 1.6.1"
gem "puma"
gem "rake"

gem 'rom', '~> 5.3'
gem 'rom-sql', '~> 3.6', '>= 3.6.1'

gem "pg"

group :development, :test do
  gem "dotenv"
end

group :cli, :development do
  gem "hanami-reloader"
end

group :cli, :development, :test do
  gem "hanami-rspec"
end

group :development do
  gem "guard-puma"
end


group :test do
  gem "rack-test"
  gem "database_cleaner-sequel"
end

```

**config/providers/persistence.rb**

```
# frozen_string_literal: true

Hanami.app.register_provider :persistence, namespace: true do
  prepare do
    require 'rom'

    config = ROM::Configuration.new(:sql, target["settings"].database_url)

    register "config", config
    register "db", config.gateways[:default].connection
  end

  start do
    config = target["persistence.config"]

    config.auto_registration(
      target.root.join('lib/startup_template/persistence'),
      namespace: 'StartupTemplate::Persistence'
    )

    register "rom", ROM.container(config)
  end
end
```

**lib/startup_template/persistence/relations/doctors.rb**

```
module StartupTemplate
  module Persistence
    module Relations
      class Doctors < ROM::Relation[:sql]
        schema(:doctors, infer: true)
      end
    end
  end
end
```

**Development Console Output**

```
startup_template$ bundle exec hanami console

startup_template[development]> StartupTemplate::Persistence
=> StartupTemplate::Persistence

startup_template[development]> StartupTemplate::Persistence::Relations
=> StartupTemplate::Persistence::Relations

startup_template[development]> StartupTemplate::Persistence::Relations::Doctors
NameError: uninitialized constant StartupTemplate::Persistence::Relations::ROM

      class Doctors < ROM::Relation[:sql]
                      ^^^
from /.../.../startup_template/lib/startup_template/persistence/relations/doctors.rb:6:in `<module:Relations>'
startup_template[development]> 
```

Exploring the source code, in my understanding it looks like `app`'s providers under `config/providers` are not getting prepared.

At following line app's providers dir location is defined

https://github.com/hanami/hanami/blob/v2.0.0/lib/hanami/slice.rb#L845

but in following method

https://github.com/hanami/hanami/blob/v2.0.0/lib/hanami/app.rb#L176-L190

that location is not used. As per `prepare_app_providers` method implementation only following providers are getting registered: `inflector`, `logger` and `rack`.

After those findings I tried following in console

```
startup_template[development]> Hanami.app.prepare(:persistence)
=> StartupTemplate::App

startup_template[development]> StartupTemplate::Persistence::Relations::Doctors
=> StartupTemplate::Persistence::Relations::Doctors

startup_template[development]> 
```

See how after preparing `:persistence` provider the NameError is resolved.

It should be noted that same error is also encountered in test env console also and how preparing the persistence provider that error resolves. Please see below

```
startup_template$ HANAMI_ENV=test bundle exec hanami console
startup_template[test]> StartupTemplate::Persistence::Relations::Doctors
NameError: uninitialized constant StartupTemplate::Persistence::Relations::ROM

      class Doctors < ROM::Relation[:sql]
                      ^^^
from /.../.../startup_template/lib/startup_template/persistence/relations/doctors.rb:6:in `<module:Relations>'

startup_template[test]> Hanami.app.prepare(:persistence)
=> StartupTemplate::App

startup_template[test]> StartupTemplate::Persistence::Relations::Doctors
=> StartupTemplate::Persistence::Relations::Doctors

startup_template[test]> 
```

Thanks.

**Comment by timriley** [comment]
Hi @jiggneshhgohel, thanks for trying Hanami 2.0!

So there's one thing that's worth knowing about these persistence classes (and this is more a rom-rb thing than a Hanami thing) — they're not designed for you to reference them directly by class name as the app author. Instead, you access them via the `rom` object that is created as part of rom-rb setup.

So given an app created by following guides (mine is called "bookshelf" and I've created a `books` relation), you can access the relation like this:

```
$ be hanami c
bookshelf[development]> app["persistence.rom"].relations[:books]
=> #<Bookshelf::Persistence::Relations::Books name=ROM::Relation::Name(books) dataset=#<Sequel::Postgres::Dataset: "SELECT \"books\".\"id\", \"books\".\"title\", \"books\".\"author\" FROM \"books\" ORDER BY \"books\".\"id\"">>
```

Better still, this is a also properly initialized relation instance (which rom takes care of as part of its setup) so you can use it for querying your database right away:

```ruby
app["persistence.rom"].relations[:books].insert(title: "POODR", author: "Sandi Metz")
app["persistence.rom"].relations[:books].to_a
=> [{:id=>1, :title=>"POODR", :author=>"Sandi Metz"}]
```

And the other nice thing about this way of accessing your relations is that you don't need to explicitly `prepare` or `start` the persistence provider. Hanami's lazy component loading takes care of that for you the first time you resolve the `"persistence.rom"` component from the app.

Hopefully this can help unblock you!

The [rom-rb](https://rom-rb.org) docs may also be helpful if you want to learn more about rom.

Given things are currently working as designed, I'm going to close this issue right now, but please do feel free to reply here if you have further questions.

**Comment by jiggneshhgohel** [comment]
@timriley Thanks for your elaborate response and esp for sharing following insights

>So there's one thing that's worth knowing about these persistence classes (and this is more a rom-rb thing than a Hanami thing) — they're not designed for you to reference them directly by class name as the app author. Instead, you access them via the rom object that is created as part of rom-rb setup.

>And the other nice thing about this way of accessing your relations is that you don't need to explicitly prepare or start the persistence provider. Hanami's lazy component loading takes care of that for you the first time you resolve the "persistence.rom" component from the app.

The way you suggested works (see console output below)

```
startup_template$ HANAMI_ENV=test bundle exec hanami c

startup_template[test]> app["persistence.rom"].relations[:doctors]
=> #<StartupTemplate::Persistence::Relations::Doctors name=ROM::Relation::Name(doctors) dataset=#<Sequel::Postgres::Dataset: "SELECT \"doctors\".\"id\", \"doctors\".\"name\", \"doctors\".\"qualification\", \"doctors\".\"created_at\", \"doctors\".\"updated_at\" FROM \"doctors\" ORDER BY \"doctors\".\"id\"">>

startup_template[test]> 
```

As a novice to hanami and also to rom-rb looking at the code examples shown under https://rom-rb.org/learn section my obvious attempt is to try out what is shown in the documentation and the Relations section at https://rom-rb.org/learn/sql/3.3/relations/ and https://rom-rb.org/learn/core/5.2/relations/ what are shown are _only classes_ so I attempted to play with Relations (through hanami console) through their classes whose definition is illustrated at https://guides.hanamirb.org/v2.0/introduction/getting-started/#persisting-books. 

Today based on your shared insights I explored the rom-rb docs more and going through data workflow at https://rom-rb.org/learn/introduction/core-concepts/#data-flow for reading and writing data I noticed that in both the cases it is mentioned to perform those operations through ROM Repositories but in Hanami 2.0 guide repositories are not demonstrated. Instead relations are illustrated to be directly accessed (like you shared) through ROM container configured through persistence provider. 

Considering above and my no experience with Hanami and rom-rb and also to start with Hanami initially following Hanami 1.3 guide to know about the concepts Hanami offers and then trying to build up a prototype web-app based on Hanami 2.0 version and trying to bring in concepts  from 1.3 -- you should be able to understand how so many things are mixed-up in a novice's mindset like me! 

When I decided to explore Hanami as an alternate to Rails, Hanami 2.0's rc-1 was available and since the 2.0 guides were in progress I assumed that following 1.3 guide should be the best option assuming that in 2.0 improvements and enhancements should be there with no changes to fundamental concepts but that assumption was wrong. 

And as per my trial with 2.0 getting started guide I ended up getting stuck with https://github.com/hanami/hanami/issues/1261#issue-1461345258. But your shared insights are very much helpful and I am hoping that it should be helpful for others like me.

Thanks again for taking time to elaborately respond to my problem.
